internal/job/checkout.go

```diff
@@ -549,9 +549,18 @@ func (e *Executor) defaultCheckoutPhase(ctx context.Context) error {
 
 	// Git clean prior to checkout, we do this even if submodules have been
 	// disabled to ensure previous submodules are cleaned up
+	gitSubmodules := false
```

Comment:
The comment above this is no longer true. I can see it was added in https://github.com/buildkite/agent/pull/1054 but there's no more information about why it was added.

Reply:
🤦🏼 I just read the comment and realised why this happened in the first place now.

Now I'm wondering what circumstances lead to this being the case, and how we could possibly cater for both situations.

Reply:
AIUI, unless it's a clean checkout (directory doesn't exist, it was empty, or the agent is configured to always do clean checkouts), the agent is running the checkout against a directory containing code that was checked out in a previous job. So it might contain submodules, that have since been removed in whatever ref is being checked out now.

So I wonder if we can harden this in some other way? Such as making `hasGitSubmodules` more robust, or skipping submodule cleaning if the checkout itself is clean?